### PR TITLE
fix: backport marktext muya editor/cursor/IME fixes (PR-3)

### PR DESCRIPTION
**Pull request by Jocs**
## Summary

Migrate marktext muya P2 (cursor / selection / autopair / IME / table /
search / history) fixes per the plan-mode tracker at
`/Users/ransixi/.claude/plans/glimmering-hatching-lightning.md`. This PR
bundles the four sub-batches the kickoff outlined (PR-3a / PR-3b /
PR-3c / PR-3d) — most marktext fixes are structurally resolved by the
new ot-json1 + block-tree architecture (no `partialRender`, no
`contentState`, fine-grained `shownFloat` gates, `composeHandler`
flipping `isComposed`), so the bulk lands as `verified-not-applicable`
with defensive tests; three items still needed real backports.

### Real backports

* **autopair (`358fa83d`, #2960)** — `Content.autoPair` now gates the
  three quote/bracket branches on `!/\\S/.test(postInputChar)` so
  `"|foo` no longer auto-expands to `""|foo`.
* **table (`5fb130d9`, #2331)** — `TableCellContent.tabHandler` branches
  on `event.shiftKey` and calls `previousContentInContext()`, enabling
  Shift+Tab back-navigation within a table; first cell of the header
  row no-ops cleanly.
* **selection (`ed1b3354`, #2816)** — image-selected keydown matcher
  expanded to `^(?:Backspace|Delete|Enter)\$`, so Delete on a selected
  image actually deletes the image instead of falling through to the
  text after it. Anchoring the regex also closes a sub-string match
  edge case.

### Coverage breakdown (per `MIGRATION.md`)

| Sub-PR | Items | Outcome |
|---|---|---|
| **PR-3a** cursor / selection | 5 | 4 verified-not-applicable + 1 test-only |
| **PR-3b** autopair | 4 | 1 fixed + 3 verified-not-applicable |
| **PR-3c** IME | 3 | 3 verified-not-applicable |
| **PR-3d** table / search / history | 11 | 2 fixed + 6 verified-not-applicable + 2 test-only + 1 deferred to PR-4 |

22 new regression tests (55 total in \`packages/core\`, up from 33).

### Items deliberately left as \`test-only\`

Two paths still depend on cross-block native browser deletion which we
can't reproduce in node-vitest:

* **\`701fb9ae\`** multi-paragraph soft-line removal (#2269)
* **\`0dc4b415\`** table cell \`<br/>X\` backspace (#1218)

Both need a real-browser pass in \`examples/\` before they can be
re-classified.

## Test plan

- [x] \`pnpm test\` — 55/55 passing (10 files)
- [x] \`pnpm lint:types\` clean
- [x] \`pnpm lint\` — 0 errors (17 pre-existing warnings untouched)
- [x] \`pnpm check-circular\` — no circular deps
- [ ] Manual \`examples/\` verification still needed for:
  - IME with multi-block selection (paragraph → another paragraph) +
    Chinese/Japanese composition (related to \`edbab6ed\`)
  - \`701fb9ae\` multi-paragraph soft-line removal
  - \`0dc4b415\` table cell \`<br/>X\` backspace
  - \`6a50b5cb\` task-list checkbox toggle cursor feel

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/core/src/block/content/tableCell/__tests__/tabHandler.spec.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
+import { describe, expect, it, vi } from 'vitest';
+import TableCellContent from '../index';
+
+// Regression for marktext 5fb130d9 "#2330 enable shift+tab for table
```
[the remark this answers is not in the document]
Comment (Jocs) [reply packages/core/src/block/content/tableCell/__tests__/tabHandler.spec.ts]: Right — marktext's commit subject is `fix: #2330 enable shift+tab for table navigation (#2331)` so the *issue* is #2330 and the *PR* is #2331. The implementation comment was using #2331 (PR) while the test header used #2330 (issue), which is confusing for archaeology. Updated both to spell out `(issue #2330, PR #2331)`. Fixed in e395d8b.

File: packages/core/src/utils/__tests__/search.spec.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+            makeMatch('2026-05-20', ['2026', '05', '20']),
+            '$3/$2/$1',
+        );
+        expect(value).toBe('20/05/2026');
+    });
+
+    it('leaves `\\$1` literal alone (escape with backslash)', () => {
+        const value = buildRegexValue(makeMatch('foo', ['cap']), 'pre \\$1 post');
+        // backslash is preserved verbatim — the regex `(?<!\\)\$\d`
+        // guards against it.
```
[the remark this answers is not in the document]
Comment (Jocs) [reply packages/core/src/utils/__tests__/search.spec.ts:39]: Two reasons I'm leaving this as-is:

1. **Chrome 70 supports lookbehind.** ECMAScript 2018 lookbehind assertions landed in V8 in Chrome 62 (Oct 2017) — see [caniuse](https://caniuse.com/js-regexp-lookbehind). Chrome 70 (Oct 2018) is well past that. The `chrome70` build target in `packages/core/vite.config.ts` is for transpilation of newer JS *syntax*, not for regex feature support.

2. **Not on the diff.** `utils/search.ts:buildRegexValue` was added in `c89e7556` (April 2024) — see `git blame` for line 37. This PR only adds defensive tests against the existing function as part of the marktext 4c517b16 migration tracker, and explicitly does not touch the implementation (it's already aligned with the marktext fix). Refactoring it would be out of scope; if lookbehind ever does become a portability concern, a separate PR can swap it for `String#replace` with a callback.

So: agreed that lookbehind is worth being mindful of, but no change needed here.

File: packages/core/src/block/base/__tests__/autoPair.spec.ts [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+// markdown-syntax branch. Defensive test.
+describe('autoPair — bbea7eca skip when preInputChar is alphanumeric', () => {
+    it('does not auto-pair `*` after a letter (typing inside a word)', () => {
+        const fakeThis = makeFakeThis('foo ', 4);
+        const event = makeInputEvent('insertText', '*');
+        const { text, needRender } = invokeAutoPair(fakeThis, event, 'foo *', 5);
+
+        // Note: text is unchanged, but the leading whitespace ensures we
+        // do not hit the `!/[a-z0-9]/i.test(preInputChar)` filter.
+        // After 'foo ' typing '*' at offset 5 → preInputChar is ' '.
```
[the remark this answers is not in the document]
Comment (Jocs) [reply packages/core/src/block/base/__tests__/autoPair.spec.ts]: Good catch — that test was actually the *positive* control (after `foo ` the preInputChar is the space, so the markdown-syntax branch *should* still fire and the closer gets inserted) but I'd left it labelled like the alnum-skip cases below it. Renamed to "still auto-pairs `*` after whitespace (the negative control)" and rewrote the comment to spell out the contrast with the two following tests. Fixed in e395d8b.
